src/shred_collector/shred_receiver.zig

```diff
@@ -0,0 +1,202 @@
+const std = @import("std");
+const network = @import("zig-network");
+const sig = @import("../lib.zig");
+
+const bincode = sig.bincode;
+const layout = sig.shred_collector.shred_layout;
+
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const Atomic = std.atomic.Value;
+const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+const Socket = network.Socket;
+
+const Channel = sig.sync.Channel;
+const Logger = sig.trace.Logger;
+const Packet = sig.net.Packet;
+const Ping = sig.gossip.Ping;
+const Pong = sig.gossip.Pong;
+const RepairMessage = sig.shred_collector.RepairMessage;
+const Slot = sig.core.Slot;
+const SocketThread = sig.net.SocketThread;
+
+const endpointToString = sig.net.endpointToString;
+
+/// Analogous to [ShredFetchStage](https://github.com/anza-xyz/agave/blob/aa2f078836434965e1a5a03af7f95c6640fe6e1e/core/src/shred_fetch_stage.rs#L34)
+pub const ShredReceiver = struct {
+    allocator: Allocator,
+    keypair: *const KeyPair,
+    exit: *Atomic(bool),
+    logger: Logger,
+    repair_socket: Socket,
+    tvu_socket: Socket,
+    outgoing_shred_channel: *Channel(ArrayList(Packet)),
```

Comment:
d8e65f73745481c2a9c03d6074c100c5e8565968

I feel torn about this because it's a form of coupling between components that are otherwise modular. It trades local intelligibility for global intelligibility. But I guess it's fine, since this is the only way these components are intended to be used at this point.

Reply:
we could add a suffix of _recv/_send to the method parameters to make it even more clear?

Reply:
That's a good point, and I'd take it one step further. I would call them senders or receivers *instead* of channels. The role of the field is to be a sender or receiver. The fact that it happens to be a channel is an implementation detail.

888747577181a4e2f78143752ae850bb86b0ed3f

Reply:
nice 🔥